File: javascript/067.js
/**
 * @param {string} a
 * @param {string} b
 * @return {string}
 */
var addBinary = function(a, b) {
  let indexA = a.length - 1;
  let indexB = b.length - 1;

  let result = "";
  let left = 0;
  while(indexA >= 0 || indexB >= 0) {
    let aValue = 0;
    let bValue = 0;
    if (indexA >= 0) aValue = parseInt(a[indexA]);
    if (indexB >= 0) bValue = parseInt(b[indexB]);
    let newValue = aValue + bValue + left;
    if (newValue >= 2) {
      left = 1;
      newValue = newValue - 2;
    } else {
      left = 0;
    }

    result = newValue + result;
    if (indexA >= 0) --indexA;
    if (indexB >= 0) --indexB;
  }

  if (left > 0) result = left + result;

  return result;
};

File: ruby/067.rb
# @param {String} a
# @param {String} b
# @return {String}
def add_binary(a, b)
  current_a = a.length - 1
  current_b = b.length - 1

  result = ""
  left = 0
  while current_a >= 0 || current_b >= 0 do
    a_value = current_a >= 0 ? a[current_a].to_i : 0
    b_value = current_b >= 0 ? b[current_b].to_i : 0
    sum = a_value + b_value + left
    if sum <= 1
      result = sum.to_s + result
      left = 0
    else
      result = (sum - 2).to_s + result
      left = 1
    end

    current_a -= 1 if current_a >= 0
    current_b -= 1 if current_b >= 0
  end

  result = left.to_s + result if left > 0

  result
end